classes/services/PKPSubmissionService.inc.php

```diff
@@ -941,4 +941,66 @@ public function updateStatus($submission)
 
         return $submission;
     }
+
+    /**
+     * Checks if this user is granted reader access to pre-publication submissions
+     * based on their roles in the context (i.e. Manager, Editor, etc).
+     *
+     * @param $user User
+     * @param $submission Submission
+     *
+     * @return bool
+     */
+    public function canPreview($user, $submission)
+    {
+        // Don't grant access until submission reaches Copyediting stage
+        if ($submission->getData('stageId') < WORKFLOW_STAGE_ID_EDITING) {
+            return false;
+        }
+
+        if ($this->_roleCanPreview($user, $submission)) {
+            return true;
+        }
+
+        if ($user) {
+            $stageAssignmentDao = DAORegistry::getDAO('StageAssignmentDAO'); /* @var $stageAssignmentDao StageAssignmentDAO */
+            $stageAssignments = $stageAssignmentDao->getBySubmissionAndRoleId($submission->getId(), ROLE_ID_AUTHOR, null, $user->getId());
+            $stageAssignment = $stageAssignments->next();
+            if ($stageAssignment) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Checks if this user is granted access to preview
+     * based on their roles in the context (i.e. Manager, Editor, etc).
+     *
+     * @param $user User
+     * @param $submission Submission
+     *
+     * @return bool
+     */
+    public function _roleCanPreview($user, $submission)
```

Comment:
Make this a `protected` function. We don't want to expose it as part of the submission service API.